monai/transforms/utility/array.py

```diff
@@ -1715,3 +1715,84 @@ def __call__(self, img: NdarrayOrTensor, meta_dict: Mapping | None = None) -> Nd
         if self._do_transform:
             img = self.filter(img)
         return img
+
+
+class CoordinateTransform(InvertibleTransform, Transform):
+    """
+    Transform points between image coordinates and world coordinates.
+
+    Args:
+        dtype: The desired data type for the output.
+        mode: Specifies the direction of transformation. This should be an instance of
+            :py:class:`CoordinateTransformMode` enum, with either 'IMAGE_TO_WORLD' or 'WORLD_TO_IMAGE' as the value.
+        affine_lps_to_ras: Defaults to ``False``. Set this to ``True`` if: 1) The image is read by ITKReader,
+            and 2) The ITKReader has `affine_lps_to_ras=True`, and 3) The data is in world coordinates.
+            This ensures that the affine transformation between LPS (left-posterior-superior) and RAS
+            (right-anterior-superior) coordinate systems is correctly applied.
+    """
+
+    def __init__(self, dtype: DtypeLike = torch.float64, mode=CoordinateTransformMode.IMAGE_TO_WORLD, affine_lps_to_ras: bool = False) -> None:
+        self.dtype = dtype
+        self.mode = mode
+        self.affine_lps_to_ras = affine_lps_to_ras
+
+    def transform_coordinates(self, data, affine, invert=False):
+        """
+        Transform coordinates using an affine transformation matrix.
+
+        Args:
+            data: The input coordinates.
+            affine: A 3x3 or 4x4 affine transformation matrix.
+            invert: Whether to invert the affine matrix.
+
+        Returns:
+            Transformed coordinates.
+        """
+        data = convert_to_tensor(data, track_meta=get_track_meta())
+        data_: torch.Tensor = convert_to_tensor(data, track_meta=False, dtype=self.dtype)
+
+        if self.affine_lps_to_ras:  # RAS affine
+            affine = orientation_ras_lps(affine)
+
+        affine_t = linalg_inv(affine)
+        _affine = affine_t if invert else affine
+
+        homogeneous = concatenate((data_, torch.ones((data_.shape[0], 1))), axis=1)
+        transformed_homogeneous = torch.matmul(_affine, homogeneous.T)
+        transformed_coordinates = transformed_homogeneous[:-1].T
+        out, *_ = convert_to_dst_type(transformed_coordinates, data, dtype=self.dtype)
+
+        extra_info = {
+            "mode": self.mode,
+            "dtype": str(self.dtype)[6:],  # dtype as string; remove "torch": torch.float32 -> float32
+            "affine": affine if invert else affine_t,
```

Comment:
This logic seems backwards to me, but I haven't parsed the full intent of this yet.

Reply:
Yeah, it may lead confusing here. I'm not convert the affine back. The affine here saved in `extra_info` is only used for inverse.
Given an example, if the transformed affine is `A`, then the affine logged here is the inverse matrix of `A` which will be used for the inverse.
I will think more and see whether I can make it clearer here.

Thanks!

Reply:
Closing this as we had already discussed it in meetings. I think @KumoLiu has already addressed this. Please feel free to re-open it if there are questions @aylward